src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -0,0 +1,644 @@
+        expr->GetChild(1)->GetExpressionType() == parser::ExpressionType::ROW_SUBQUERY) {
+      throw NOT_IMPLEMENTED_EXCEPTION("Do not support comparison between sub-select");
+    }
+    // Transform if either child is sub-query
+    GenerateSubqueryTree(expr, 0, parse_result, true) || GenerateSubqueryTree(expr, 1, parse_result, true);
+  }
+  expr->AcceptChildren(this, parse_result);
+}
+
+void QueryToOperatorTransformer::Visit(parser::OperatorExpression *expr, parser::ParseResult *parse_result) {
+  // TODO(boweic): We may want to do the rewrite (exist -> in) in the binder
+  if (expr->GetExpressionType() == parser::ExpressionType::OPERATOR_EXISTS) {
+    if (GenerateSubqueryTree(expr, 0, parse_result, false)) {
+      // Already reset the child to column, we need to transform exist to not-null to preserve semantic
+      expr->SetExpressionType(parser::ExpressionType::OPERATOR_IS_NOT_NULL);
+    }
+  }
+
+  expr->AcceptChildren(this, parse_result);
+}
+
+bool QueryToOperatorTransformer::RequireAggregation(common::ManagedPointer<parser::SelectStatement> op) {
+  if (op->GetSelectGroupBy() != nullptr) {
+    return true;
+  }
+  // Check plain aggregation
+  bool has_aggregation = false;
+  bool has_other_exprs = false;
+
+  for (auto &expr : op->GetSelectColumns()) {
+    std::vector<common::ManagedPointer<parser::AggregateExpression>> aggr_exprs;
+    // we need to use get method of managed pointer because the function we are calling will recursivly get aggreate
+    // expressions from the current expression and its children; children are of unique pointers
+    parser::expression::ExpressionUtil::GetAggregateExprs(&aggr_exprs, expr);
+    if (!aggr_exprs.empty())
+      has_aggregation = true;
+    else
+      has_other_exprs = true;
+  }
+  // TODO(peloton): Should be handled in the binder
+  // Syntax error when there are mixture of aggregation and other exprs when group by is absent
+  if (has_aggregation && has_other_exprs) {
+    throw OPTIMIZER_EXCEPTION(
+        "Non aggregation expression must appear in the GROUP BY clause or be used in an aggregate function");
+  }
+  return has_aggregation;
+}
+
+void QueryToOperatorTransformer::CollectPredicates(common::ManagedPointer<parser::AbstractExpression> expr,
+                                                   parser::ParseResult *parse_result,
+                                                   std::vector<AnnotatedExpression> *predicates) {
+  // First check if all conjunctive predicates are supported before transforming
+  // predicate with sub-select into regular predicates
+  std::vector<common::ManagedPointer<parser::AbstractExpression>> predicate_ptrs;
+  QueryToOperatorTransformer::SplitPredicates(expr, &predicate_ptrs);
+  for (const auto &pred : predicate_ptrs) {
+    if (!QueryToOperatorTransformer::IsSupportedConjunctivePredicate(pred)) {
+      throw NOT_IMPLEMENTED_EXCEPTION("Predicate type not supported yet");
```

Comment:
Rephrase error message: `"Expression " + pred->GetExpr()->GetExpressionType() + " is not supported"`